### PR TITLE
API Refactor Form request handling into FormRequestHandler

### PR DESCRIPTION
**Pull request by tractorcow**
Fixes https://github.com/silverstripe/silverstripe-framework/issues/5639
Fixes https://github.com/silverstripe/silverstripe-framework/issues/6362

Requires https://github.com/silverstripe/silverstripe-cms/pull/1758

### CODE REVIEW

**Comment by dhensby** [comment]
There are some CS failures

File: src/Control/RequestHandler.php [hunk 1]
```diff
@@ -546,4 +547,114 @@ public function setRequest($request)
...
+    {
+        $request = $this->getRequest();
+        if (!$request) {
+            return null;
+        }
+        $pageURL = $request->getHeader('Referer');
+        if (!$pageURL) {
+            return null;
+        }
+        if (Director::is_site_url($pageURL)) {
```
Comment (dhensby) [inline src/Control/RequestHandler.php]: Why should the `getReferer` be opinionated about whether the url is a site url or not?

File: src/Security/Security.php [hunk 2]
```diff
@@ -2,17 +2,15 @@
 
 namespace SilverStripe\Security;
 
+use Page;
```
Comment (dhensby) [inline src/Security/Security.php]: `Page` is never used in this class (apart from for a phpdoc type hint)

File: src/Control/RequestHandler.php (hunk 1 above)
Comment (tractorcow) [inline src/Control/RequestHandler.php]: Well, I can call it `getInternalReferer`, but the intent of this method is to get only a referer that's safe to redirect to.

I could add a flag to it to allow external urls to be allowed?

Comment (tractorcow) [inline src/Control/RequestHandler.php]:
```
/**
 * Get referrer to redirect back to and safely validates it
 */
```

File: src/Security/Security.php (hunk 2 above)
Comment (tractorcow) [inline src/Security/Security.php]: I'll change this reference to a class reference.
`private static $page_class = Page::class;`

File: src/Control/RequestHandler.php (hunk 1 above)
Comment (tractorcow) [inline src/Control/RequestHandler.php]: I'll split it into getReferer() (raw) and getReturnReferer() (safely validates)

**Comment by tractorcow** [comment]
Fixed up linting issues, split getReferer() into getReferer() and getReturnReferer(), and replaced a string literal with class constant in Security.php.

File: src/Control/RequestHandler.php (hunk 1 above)
Comment (dhensby) [inline src/Control/RequestHandler.php]: Maybe `getSafeReferer` is all we need as the code to get the referer header is a one liner anyway?

**Comment by dhensby** [comment]
The CMS PR needs fixing before we can merge

**Comment by tractorcow** [comment]
Fixed @dhensby

**Pull request merged**